src/compiler/emitter.ts

```diff
@@ -1179,4394 +1178,5531 @@ namespace ts {
-                        if (seenPrologueDirectives) {
-                            seenPrologueDirectives.add(statement.expression.text);
-                        }
-                    }
+        function writeTokenText(token: SyntaxKind, writer: (s: string) => void): void;
+        function writeTokenText(token: SyntaxKind, writer: (s: string) => void, pos: number): number;
+        function writeTokenText(token: SyntaxKind, writer: (s: string) => void, pos?: number): number {
+            const tokenString = tokenToString(token)!;
+            writer(tokenString);
+            return pos! < 0 ? pos! : pos! + tokenString.length;
+        }
+
+        function writeLineOrSpace(parentNode: Node, prevChildNode: Node, nextChildNode: Node) {
+            if (getEmitFlags(parentNode) & EmitFlags.SingleLine) {
+                writeSpace();
+            }
+            else if (preserveSourceNewlines) {
+                const lines = getLinesBetweenNodes(parentNode, prevChildNode, nextChildNode);
+                if (lines) {
+                    writeLine(lines);
                 }
                 else {
-                    // return index of the first non prologue directive
-                    return i;
+                    writeSpace();
                 }
             }
-
-            return statements.length;
+            else {
+                writeLine();
+            }
         }
 
-        function emitUnparsedPrologues(prologues: readonly UnparsedPrologue[], seenPrologueDirectives: Set<string>) {
-            for (const prologue of prologues) {
-                if (!seenPrologueDirectives.has(prologue.data)) {
+        function writeLines(text: string): void {
+            const lines = text.split(/\r\n?|\n/g);
+            const indentation = guessIndentation(lines);
+            for (const lineText of lines) {
+                const line = indentation ? lineText.slice(indentation) : lineText;
+                if (line.length) {
                     writeLine();
-                    const pos = writer.getTextPos();
-                    emit(prologue);
-                    if (bundleFileInfo) bundleFileInfo.sections.push({ pos, end: writer.getTextPos(), kind: BundleFileSectionKind.Prologue, data: prologue.data });
-                    if (seenPrologueDirectives) {
-                        seenPrologueDirectives.add(prologue.data);
-                    }
+                    write(line);
                 }
             }
         }
 
-        function emitPrologueDirectivesIfNeeded(sourceFileOrBundle: Bundle | SourceFile) {
-            if (isSourceFile(sourceFileOrBundle)) {
-                emitPrologueDirectives(sourceFileOrBundle.statements, sourceFileOrBundle);
+        function writeLinesAndIndent(lineCount: number, writeSpaceIfNotIndenting: bo
```

Comment:
This probably isn't relevant to this PR, but I'm wondering if at this point the `BinaryExpression` trampoline pattern is worth extracting into some kind of common factory - very similar trampolines exist in the preprinter, emitter, checker, and, IIRC, binder now.